File: iosApp/iosApp/Services/CurrencyStoreServiceAssembler.swift
// Created by web3d3v on 31/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

final class CurrencyStoreServiceAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .singleton) { resolver -> CurrencyStoreService in
            let service = DefaultCurrencyStoreService(
                coinGeckoService: DefaultCoinGeckoService(),
                marketStore: KeyValueStore(name: "CurrencyStoreService.Market"),
                candleStore: KeyValueStore(name: "CurrencyStoreService.Candle"),
                metadataStore: KeyValueStore(name: "CurrencyStoreService.Metadata"),
                userCurrencyStore: KeyValueStore(
                    name: "CurrencyStoreService.UserCurrency"
                )
            )
            service.loadCaches(
                networks: NetworksServiceCompanion().supportedNetworks(),
                completionHandler: { _, _ in () }
            )
            return service
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/services/CurrencyStoreServiceAssembler.kt
package com.sonsofcrypto.web3wallet.android.services

import com.sonsofcrypto.web3lib.keyValueStore.KeyValueStore
import com.sonsofcrypto.web3lib.services.coinGecko.CoinGeckoService
import com.sonsofcrypto.web3lib.services.coinGecko.DefaultCoinGeckoService
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.currencyStore.DefaultCurrencyStoreService
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import smartadapter.internal.extension.name

class CurrencyStoreServiceAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(CurrencyStoreService::class.name, AssemblerRegistryScope.SINGLETON) {
            DefaultCurrencyStoreService(
                it.resolve("${CoinGeckoService::class.name}"),
                it.resolve("${KeyValueStore::class.name}.Market"),
                it.resolve("${KeyValueStore::class.name}.Candle"),
                it.resolve("${KeyValueStore::class.name}.Metadata"),
                it.resolve("${KeyValueStore::class.name}.UserCurrency"),
            )
            // TODO: Review loading Cache...should not be done by the assembler
//            service.loadCaches(
//                NetworksService.Companion.supportedNetworks(),
//            )
            // return service
        }
    }
}